pkg/apis/networking/v1alpha1/domain_types.go

```diff
@@ -0,0 +1,121 @@
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the Domain.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec DomainSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the Domain.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status DomainStatus `json:"status,omitempty"`
+}
+
+// Verify that Domain adheres to the appropriate interfaces.
+var (
+	// Check that Doimain may be validated and defaulted.
+	_ apis.Validatable = (*Domain)(nil)
+	_ apis.Defaultable = (*Domain)(nil)
+
+	// Check that we can create OwnerReferences to a Domain.
+	_ kmeta.OwnerRefable = (*Domain)(nil)
+
+	// Check that the type conforms to the duck Knative Resource shape.
+	_ duckv1.KRShaped = (*Domain)(nil)
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DomainList is a collection of Domain objects.
+type DomainList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of Domain objects.
+	Items []Domain `json:"items"`
+}
+
+// DomainSpec describes the Ingress the user wishes to exist.
+//
+// In general this follows the same shape as K8s Ingress.
+// Some notable differences:
+// - Backends now can have namespace:
+// - Traffic can be split across multiple backends.
+// - Timeout & Retry can be configured.
+// - Headers can be appended.
+// DomainSpec contains the specification of the Domain CRD.
+type DomainSpec struct {
+	// IngressClass tells what Ingress class annotation to use for Routes associated
+	// with this Realm.
+	IngressClass string `json:"ingressClass"`
+
+	// Suffix specifies the domain suffix to be used. This field replaces the
+	// existing config-domain ConfigMap.  Internal Domains can omit this, in
+	// which case we will default to the cluster suffix.
+	// +optional
+	Suffix string `json:"suffix,omitempty"`
+
+	// LoadBalancers provide addresses (IP address, domains) of the load balancers
+	// associated with this Domain.  This is used in automatic DNS provisioning like
+	// configuration of magic DNS or creating ExternalName services for cluster-local
+	// access.
+	LoadBalancers []LoadBalancerIngressStatus `json:"loadBalancers"`
+
+	// Config allows KIngress implementations to add additional information needed
+	// for configuring the proxies associated with this Domain.
+	// For examples, in our Istio-based Ingress this will contains all references of
+	// Istio Gateways associated with this Domain. This could be a reference of a ConfigMap
+	// owned by the implementation as well.
+	// +optional
+	Config []map[string]string `json:"config,omitempty"`
```

Comment:
Some people think we should not use `maps` in the apis
/cc @mattmoor

Reply:
cc @dprotaso can you please suggest the best practice to represent maps here?

Reply:
This is an array of maps - is that intended?

> best practice to represent maps here?

Maybe a few criteria to think about

### Patching
Although, CRDs don't support strategic merge patches the different types help reason about the different patch strategies that could open up in the future (although it's been shut down a bunch of times)

https://pkg.go.dev/k8s.io/kube-openapi/pkg/idl?tab=doc#ListType
https://pkg.go.dev/k8s.io/kube-openapi/pkg/idl?tab=doc#MapType
https://pkg.go.dev/k8s.io/kube-openapi/pkg/idl?tab=doc#StructType

[JSON Patch](https://tools.ietf.org/html/rfc6902) & [JSON Merge Patch](https://tools.ietf.org/html/rfc7386) are applicable to CRDs.


### Future extensibility 
Right now using an array with maps above don't let you add new typed fields - this may be fine but something to consider.

ie. https://github.com/knative/serving/pull/7373#discussion_r407716556

Reply:
For conditions there's also 
https://github.com/kubernetes/community/pull/4521#issuecomment-627365318

Reply:
Thanks for detailed explanation @dprotaso. Updated the type to use array.